hs_core/views/__init__.py

```diff
@@ -313,17 +342,39 @@ def publish(request, shortkey, *args, **kwargs):
 def set_resource_flag(request, shortkey, *args, **kwargs):
     # only resource owners are allowed to change resource flags
     res, _, user = authorize(request, shortkey, needed_permission=ACTION_TO_AUTHORIZE.SET_RESOURCE_FLAG)
-    t = request.POST['t']
+    if 't' in kwargs:
+        t = kwargs['t']
+    else:
+        t = request.POST['t']
     if t == 'make_public':
         _set_resource_sharing_status(request, user, res, flag_to_set='public', flag_value=True)
     elif t == 'make_private' or t == 'make_not_discoverable':
+
         _set_resource_sharing_status(request, user, res, flag_to_set='public', flag_value=False)
+
+        # downgrade all public collections that hold this res to private (because public collections can hold non-private resources only)
+        # reverse lookup: find all collections that hold this res being set flag
+        associated_collection_metadata_obj_list = list(res.collection_set.all())
```

Comment:
This is the same code you have added also in the ```delete_resource()``` view function. Should put this common code in a new function and call that function in both places.